feat(components): [Carousel] Add motion blur functionality

JedediahXu opened the pull request:
* docs(components): [Carousel] Add example properties

* docs(components): [Carousel] Add docs properties

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

# Why add motion blur  为什么要添加运动模糊

Add motion blur for a better user experience 

添加运动模糊以获得更好的用户体验

## 特性

1、Smooth Transition Effect: Motion blur can create a smooth transition effect in the carousel, making the images appear more natural during switches and reducing the abruptness associated with sudden transitions.
     平滑过渡效果: 运动模糊可以在轮播中创建一个平滑的过渡效果，使图像在切换时看起来更自然，减少突然过渡带来的突兀感。

2、Highlighting Focus: Emphasizing the primary content within the current carousel slide, making it more prominent during transitions.
    突出焦点: 强调当前轮播幻灯片中的主要内容，使其在过渡期间更加突出。

3、Dynamics and Fluidity: Introducing a sense of dynamism, providing users with a smoother and captivating visual experience, thereby enhancing overall appeal.
     动态和流体感: 增加了动态感，为用户提供了更流畅和引人入胜的视觉体验，从而增强整体吸引力。

4、Modernity and Artistry: Infused a touch of modernity and artistic flair to the carousel, giving the page a creative and stylish appearance.
     现代感和艺术感: 为轮播增添了一丝现代感和艺术气息，让页面具有创意和时尚的外观。

5、Reducing Eye Fatigue: Smooth motion effects help alleviate potential eye strain for users when viewing carousels.
     减少眼睛疲劳: 平滑的运动效果有助于减少用户在观看轮播时可能产生的眼睛疲劳 

Additionally, it seems that this feature is not yet available in other Vue component libraries.
此外，这个特性在其他Vue组件库中似乎还不可用。

##  Any Break Changes 是否有破坏性

No destructive update
无破坏性更新

## discuss 讨论

#15399

JedediahXu committed:
feat(components): [Carousel] Add motion blur functionality

docs(components): [Carousel] Add example properties

docs(components): [Carousel] Add docs properties

JedediahXu commented:
Sorry for the inconvenience. In the previous pull request, I did not follow the guidelines for writing Git commit messages in a certain part. Therefore, I have closed the previous PR. I have submitted a new PR with the corrected commit message.

![image](https://github.com/element-plus/element-plus/assets/44160015/5e281ddf-385f-4d04-b76c-dcc0ff77011d)

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

btea commented:
cc @element-plus/backers

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

FrontEndDog commented on packages/components/carousel/src/use-carousel.ts:
```diff
@@ -46,6 +46,8 @@ export const useCarousel = (
   const root = ref<HTMLDivElement>()
   const containerHeight = ref<number>(0)
   const isItemsTwoLength = ref(true)
+  const isFirstCall = ref(true)
+  const isTransitioning = ref(false)
```
why not use ‘@transitionstart’ instead ‘isFirstCall’

JedediahXu replied:
I feel using isFirstCall might simplify the code since it eliminates the need for handling event listeners and event handler functions.

FrontEndDog reviewed (changes requested):
![image](https://github.com/element-plus/element-plus/assets/46493087/7e293d4e-6bfa-47b9-92bb-7c04f119158e)

Arrows should not become blurry.

JedediahXu commented on packages/components/carousel/src/carousel.vue:
```diff
@@ -65,6 +69,21 @@
+      class="filters"
+      style="display: none"
+    >
+      <defs>
+        <filter id="horizontal">
+          <feGaussianBlur in="SourceGraphic" stdDeviation="12,0" />
+        </filter>
+        <filter id="vertical">
+          <feGaussianBlur in="SourceGraphic" stdDeviation="0,10" />
+        </filter>
+      </defs>
```
Ok, I'll change that

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
fix(components): [Carousel] change the id name in filter in svg

fix(components): [Carousel] button to remove motion blur effect

fix(components): [Carousel] svg hides when motion blur is false

JedediahXu commented:
> ![image](https://private-user-images.githubusercontent.com/46493087/302083630-7e293d4e-6bfa-47b9-92bb-7c04f119158e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwNDY1MjUsIm5iZiI6MTcwNzA0NjIyNSwicGF0aCI6Ii80NjQ5MzA4Ny8zMDIwODM2MzAtN2UyOTNkNGUtNmJmYS00N2I5LTkyYmItN2MwNGYxMTkxNThlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA0VDExMzAyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwZjAzNzdjMDE4NTUyNmIxNzYxZjMzOWM5MThjYmQ3YjFkNDg4MDBmYjA4MGMzMDg4YjA4YzAwMjViODE0YTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rFklyNnEDbflI4ZZH1RmXdvfJaBX2buktPOKMsCh_8M)
> 
> Arrows should not become blurry.

To eliminate the motion blur effect on the button, I made some slight adjustments to the layout of the button's div. Please take a look when you have a moment. Thank you for your assistance.

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu commented:
@FrontEndDog  Hey hey, if you have time, could you please help me review this? Thank you.

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

FrontEndDog reviewed (approved):
Thank you for your contribute!

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu commented:
@kooriookami Hi, if you have time, could you please help me review this? Thank you.

kooriookami commented on docs/en-US/component/carousel.md:
```diff
@@ -15,6 +15,16 @@ carousel/basic
 
 :::
 
+## Motion blur
```
```suggestion
## Motion blur ^(2.6.0)
```

kooriookami commented on docs/en-US/component/carousel.md:
```diff
@@ -80,6 +90,7 @@ carousel/vertical
 | loop               | display the items in loop                             | boolean | -                   | true       |
 | direction          | display direction                                     | string  | horizontal/vertical | horizontal |
 | pause-on-hover     | pause autoplay when hover                             | boolean | -                   | true       |
+| motion-blur        | infuse dynamism and smoothness into the carousel      | boolean | -                   | false      |
```
please add a tag version of `2.6.0`.

kooriookami commented on docs/examples/carousel/motion-blur.vue:
```diff
@@ -0,0 +1,43 @@
+<template>
+  <div class="block text-center">
+    <span class="demonstration">Motion blur the switch (default)</span>
+    <el-carousel height="200px" :motion-blur="true">
```
```suggestion
    <el-carousel height="200px" motion-blur>
```

kooriookami commented on docs/examples/carousel/motion-blur.vue:
```diff
@@ -0,0 +1,43 @@
+  <el-carousel
+    height="200px"
+    direction="vertical"
+    :motion-blur="true"
```
```suggestion
    motion-blur
```

kooriookami reviewed (changes requested):
When starting blur, the carousel item will shake. Is that a problem?

https://github.com/element-plus/element-plus/assets/38392315/6485f476-ebeb-44e4-bd12-0a218e80d73d

JedediahXu committed:
update(components): [Carousel]  docs/en-US/component/carousel.md

Co-authored-by: kooriookami <38392315+kooriookami@users.noreply.github.com>

JedediahXu committed:
update(components): [Carousel] docs/examples/carousel/motion-blur.vue

Co-authored-by: kooriookami <38392315+kooriookami@users.noreply.github.com>

JedediahXu committed:
update(components): [Carousel] docs/examples/carousel/motion-blur.vue

Co-authored-by: kooriookami <38392315+kooriookami@users.noreply.github.com>

JedediahXu committed:
update(components): [Carousel] packages/components/carousel/src/carousel.ts

Co-authored-by: kooriookami <38392315+kooriookami@users.noreply.github.com>

JedediahXu commented:
> When starting blur, the carousel item will shake. Is that a problem?
> 
>  QQ2024228-21171-HD.mp4

Thank you for your review.  This is not an issue. When the carousel scrolls, an SVG mask is added to it. This creates a motion blur effect as a result of the visual effect.

JedediahXu committed:
docs(components): [Carousel]  add feature version labels

kooriookami reviewed (approved):
LGTM

kooriookami commented:
Have you finished?

JedediahXu committed:
Merge branch 'dev' into dev

JedediahXu commented:
> Have you finished?

I have completed it @kooriookami

kooriookami commented:
Thank you for your contribution.
